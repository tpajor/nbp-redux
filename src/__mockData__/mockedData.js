const card = { id: 'someId', code: 'USD', currency: 'dolar amerykański', rates: [{ ask: 3.3, bid: 3.4 }]};
const card2 = { id: 'someId2', code: 'USD', currency: 'dolar amerykański', rates: { ask: 3.3, bid: 3.4 }};

const search = {
  cards: [card, card2],
  temporaryCard: card,
  searchError: false,
  addCardError: true,
  addCardErrorMessage: 'error',
  currenciesTable: ['usd', 'chf', 'aud'],
  currencyCodeToViewInDetail: 'usd',
};

const user = { userName: 'User', pass: 'pass' };
const user2 = { userName: 'User2', pass: 'pass2'};

export const login = {
  signingError: true,
  signingErrorMessage: 'error',
  signedIn: true,
  userSignedIn: 'User',
  users: [user, user2]
};

const lastRates = [
  { ask: 3.5983, bid: 3.5271, effectiveDate: "2018-05-07", no: "087/C/NBP/2018" },
  { ask: 3.6088, bid: 3.5374, effectiveDate: "2018-05-08", no: "088/C/NBP/2018" }
];

const detail = {
  lastRates: lastRates,
  detailsInputError: false,
  detailsInputErrorMessage: '',
};

export const initialState = { search, login, detail };

export const searchViewData = {
  cards: search.cards,
  temporaryCard: search.temporaryCard,
  searchError: search.searchError,
  signedIn: login.signedIn,
  userSignedIn: login.userSignedIn,
  addCardError: search.addCardError,
  addCardErrorMessage: search.addCardErrorMessage,
  currenciesTable: search.currenciesTable,
};

export const detailViewData = {
  currencyCodeToViewInDetail: search.currencyCodeToViewInDetail,
  lastRates: detail.lastRates,
  detailsInputError: detail.detailsInputError,
  detailsInputErrorMessage: detail.detailsInputErrorMessage,
  signedIn: login.signedIn,
};