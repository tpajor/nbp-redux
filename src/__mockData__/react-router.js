const Link = () => {
  return jest.fn();
}

export default Link;