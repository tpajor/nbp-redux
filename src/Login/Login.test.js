import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import LoginConnected, { Login } from './Login';
import { login, initialState } from '../__mockData__/mockedData';
import configureMockStore from 'redux-mock-store';

const mockStore = configureMockStore();

Enzyme.configure({ adapter: new Adapter() });

const initialComponentState = { 
  userName: '',
  pass: '',
  repeatPass: '',
  signUp: false,
};

const customState = { 
  userName: 'a',
  pass: 'b',
  repeatPass: 'b',
  signUp: true,
};

function setup() {
  const props = {
    signingError: false,
    signingErrorMessage: '',
    signedIn: false,
    users: login.users,
    signInOrOut: jest.fn(),
    registerUser: jest.fn(),
  };

  const enzymeWrapper = shallow(<Login {...props} />);

  return {
    enzymeWrapper,
    props
  };
};

describe('Login component', () => {
  it('should render self and subcomponents with their props', () => {
    const { enzymeWrapper, props } = setup();

    expect(enzymeWrapper.length).toBe(1);

    expect(enzymeWrapper.find('div').length).toBe(5);
    enzymeWrapper.setState({ ...initialComponentState, signUp: true });
    expect(enzymeWrapper.find('div').length).toBe(6);
    expect(enzymeWrapper.find('div').first().hasClass('LoginForm')).toBe(true);
    expect(enzymeWrapper.find('div').at(1).hasClass('form-group')).toBe(true);
    expect(enzymeWrapper.find('div').at(2).hasClass('form-group')).toBe(true);
    expect(enzymeWrapper.find('div').at(3).hasClass('form-group')).toBe(true);
    expect(enzymeWrapper.find('div').at(4).hasClass('form-group')).toBe(true);
    expect(enzymeWrapper.find('div').last().hasClass('form-group')).toBe(true);
    enzymeWrapper.setState({ ...initialComponentState });

    expect(enzymeWrapper.find('input').length).toBe(2);
    enzymeWrapper.setState({ ...initialComponentState, signUp: true });
    expect(enzymeWrapper.find('input').length).toBe(3);
    expect(enzymeWrapper.find('input').first().hasClass('form-control ')).toBe(true);
    expect(enzymeWrapper.find('input').at(1).hasClass('form-control ')).toBe(true);
    expect(enzymeWrapper.find('input').last().hasClass('form-control ')).toBe(true);
    enzymeWrapper.setProps({ ...props, signingError: true });
    expect(enzymeWrapper.find('input').first().hasClass('form-control is-invalid')).toBe(true);
    expect(enzymeWrapper.find('input').at(1).hasClass('form-control is-invalid')).toBe(true);
    expect(enzymeWrapper.find('input').last().hasClass('form-control is-invalid')).toBe(true);
    enzymeWrapper.setState({ ...customState });
    expect(enzymeWrapper.find('input').first().props().value).toEqual('a');
    expect(enzymeWrapper.find('input').at(1).props().value).toEqual('b');
    expect(enzymeWrapper.find('input').last().props().value).toEqual('b');
    enzymeWrapper.setState({ ...initialComponentState });
    enzymeWrapper.setProps({ ...props });

    expect(enzymeWrapper.find('p').length).toBe(0);
    enzymeWrapper.setProps({ ...props, signingError: true, signingErrorMessage: 'error' });
    expect(enzymeWrapper.find('p').length).toBe(1);
    expect(enzymeWrapper.find('p').text()).toEqual('error');
    enzymeWrapper.setProps({ ...props });

    expect(enzymeWrapper.find('button').length).toBe(2);
    expect(enzymeWrapper.find('button').first().text()).toEqual('Zaloguj');
    expect(enzymeWrapper.find('button').last().text()).toEqual('Zarejestruj sie');
    enzymeWrapper.setState({ ...initialComponentState, signUp: true });
    expect(enzymeWrapper.find('button').first().text()).toEqual('Zarejestruj');
    expect(enzymeWrapper.find('button').last().text()).toEqual('Logowanie');

    enzymeWrapper.setProps({ ...props, signedIn: true });
    expect(enzymeWrapper.find('.LoginForm').length).toBe(0);
  });

  it('should render correctly', () => {
    const { enzymeWrapper } = setup();
    
    expect(enzymeWrapper).toMatchSnapshot();
  });

  it('should set state on change on inputs', () => {
    const { enzymeWrapper } = setup();
    let input = enzymeWrapper.find('input').first();

    expect(enzymeWrapper.state(['userName'])).toEqual('');
    input.simulate('change', { target: { value: 'a' } });
    expect(enzymeWrapper.state(['userName'])).toEqual('a');
    enzymeWrapper.setState({ ...initialState });

    input = enzymeWrapper.find('input').last();
    expect(enzymeWrapper.state(['pass'])).toEqual('');
    input.simulate('change', { target: { value: 'a' } });
    expect(enzymeWrapper.state(['pass'])).toEqual('a');
    enzymeWrapper.setState({ ...initialState, signUp: true });

    input = enzymeWrapper.find('input').last();
    expect(enzymeWrapper.state(['repeatPass'])).toEqual('');
    input.simulate('change', { target: { value: 'a' } });
    expect(enzymeWrapper.state(['repeatPass'])).toEqual('a');
  });

  it('should call proper function on submit', () => {
    const { enzymeWrapper, props } = setup();
    const form = enzymeWrapper.find('form');
    
    enzymeWrapper.setState({ ...customState, signUp: false });
    form.simulate('submit', { preventDefault: () => {} });
    expect(props.signInOrOut.mock.calls.length).toBe(1);
    expect(props.signInOrOut).toHaveBeenCalledWith(false, 'a', 'b', props.users);

    enzymeWrapper.setState({ ...customState });
    form.simulate('submit', { preventDefault: () => {} });
    expect(props.registerUser.mock.calls.length).toBe(1);
    expect(props.registerUser).toHaveBeenCalledWith('a', 'b', 'b', props.users);
  });

  it('should set signUp flag in state on second button click', () => {
    const { enzymeWrapper } = setup();
    const button = enzymeWrapper.find('button').last();
    
    expect(enzymeWrapper.state(['signUp'])).toBe(false);
    button.simulate('click');
    expect(enzymeWrapper.state(['signUp'])).toBe(true);
    button.simulate('click');
    expect(enzymeWrapper.state(['signUp'])).toBe(false);
  });

  it('should map state and dispatch to props', () => {
    const store = mockStore(initialState);
    const enzymeWrapper = shallow(<LoginConnected store={store} />);
    expect(enzymeWrapper.props().signingError).toBe(true);
    expect(enzymeWrapper.props().signingErrorMessage).toEqual('error');
    expect(enzymeWrapper.props().signedIn).toBe(true);
    expect(enzymeWrapper.props().users).toEqual(login.users);
    expect(typeof enzymeWrapper.props().signInOrOut).toBe('function');
    expect(typeof enzymeWrapper.props().registerUser).toBe('function');
  });
});