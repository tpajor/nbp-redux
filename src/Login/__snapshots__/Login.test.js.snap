// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    registerUser={[Function]}
    signInOrOut={[Function]}
    signedIn={false}
    signingError={false}
    signingErrorMessage=""
    users={
        Array [
            Object {
              "pass": "pass",
              "userName": "User",
            },
            Object {
              "pass": "pass2",
              "userName": "User2",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          className="LinkToNbp"
          replace={false}
          to="/"
>
          Przejdź do kursów walut bez logowania
</Link>,
        <form
          onSubmit={[Function]}
>
          <div
                    className="form-group"
          >
                    <input
                              className="form-control "
                              onChange={[Function]}
                              placeholder="Użytkownik"
                              type="text"
                              value=""
                    />
                    
          </div>
          <div
                    className="form-group"
          >
                    <input
                              className="form-control "
                              onChange={[Function]}
                              placeholder="Hasło"
                              type="password"
                              value=""
                    />
          </div>
          
          <div
                    className="form-group"
          >
                    <button
                              className="btn btn-primary LoginButton"
                              type="submit"
                    >
                              Zaloguj
                    </button>
          </div>
</form>,
        <div
          className="form-group"
>
          <button
                    className="btn btn-info"
                    onClick={[Function]}
          >
                    Zarejestruj sie
          </button>
</div>,
      ],
      "className": "LoginForm",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Przejdź do kursów walut bez logowania",
          "className": "LinkToNbp",
          "replace": false,
          "to": "/",
        },
        "ref": null,
        "rendered": "Przejdź do kursów walut bez logowania",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="form-group"
>
              <input
                            className="form-control "
                            onChange={[Function]}
                            placeholder="Użytkownik"
                            type="text"
                            value=""
              />
              
</div>,
            <div
              className="form-group"
>
              <input
                            className="form-control "
                            onChange={[Function]}
                            placeholder="Hasło"
                            type="password"
                            value=""
              />
</div>,
            "",
            <div
              className="form-group"
>
              <button
                            className="btn btn-primary LoginButton"
                            type="submit"
              >
                            Zaloguj
              </button>
</div>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  className="form-control "
                  onChange={[Function]}
                  placeholder="Użytkownik"
                  type="text"
                  value=""
/>,
                "",
              ],
              "className": "form-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "form-control ",
                  "onChange": [Function],
                  "placeholder": "Użytkownik",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "",
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="form-control "
                onChange={[Function]}
                placeholder="Hasło"
                type="password"
                value=""
/>,
              "className": "form-group",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control ",
                "onChange": [Function],
                "placeholder": "Hasło",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
          "",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <button
                className="btn btn-primary LoginButton"
                type="submit"
>
                Zaloguj
</button>,
              "className": "form-group",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Zaloguj",
                "className": "btn btn-primary LoginButton",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Zaloguj",
              "type": "button",
            },
            "type": "div",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="btn btn-info"
            onClick={[Function]}
>
            Zarejestruj sie
</button>,
          "className": "form-group",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Zarejestruj sie",
            "className": "btn btn-info",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Zarejestruj sie",
          "type": "button",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            className="LinkToNbp"
            replace={false}
            to="/"
>
            Przejdź do kursów walut bez logowania
</Link>,
          <form
            onSubmit={[Function]}
>
            <div
                        className="form-group"
            >
                        <input
                                    className="form-control "
                                    onChange={[Function]}
                                    placeholder="Użytkownik"
                                    type="text"
                                    value=""
                        />
                        
            </div>
            <div
                        className="form-group"
            >
                        <input
                                    className="form-control "
                                    onChange={[Function]}
                                    placeholder="Hasło"
                                    type="password"
                                    value=""
                        />
            </div>
            
            <div
                        className="form-group"
            >
                        <button
                                    className="btn btn-primary LoginButton"
                                    type="submit"
                        >
                                    Zaloguj
                        </button>
            </div>
</form>,
          <div
            className="form-group"
>
            <button
                        className="btn btn-info"
                        onClick={[Function]}
            >
                        Zarejestruj sie
            </button>
</div>,
        ],
        "className": "LoginForm",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Przejdź do kursów walut bez logowania",
            "className": "LinkToNbp",
            "replace": false,
            "to": "/",
          },
          "ref": null,
          "rendered": "Przejdź do kursów walut bez logowania",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="form-group"
>
                <input
                                className="form-control "
                                onChange={[Function]}
                                placeholder="Użytkownik"
                                type="text"
                                value=""
                />
                
</div>,
              <div
                className="form-group"
>
                <input
                                className="form-control "
                                onChange={[Function]}
                                placeholder="Hasło"
                                type="password"
                                value=""
                />
</div>,
              "",
              <div
                className="form-group"
>
                <button
                                className="btn btn-primary LoginButton"
                                type="submit"
                >
                                Zaloguj
                </button>
</div>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    className="form-control "
                    onChange={[Function]}
                    placeholder="Użytkownik"
                    type="text"
                    value=""
/>,
                  "",
                ],
                "className": "form-group",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "form-control ",
                    "onChange": [Function],
                    "placeholder": "Użytkownik",
                    "type": "text",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                "",
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  className="form-control "
                  onChange={[Function]}
                  placeholder="Hasło"
                  type="password"
                  value=""
/>,
                "className": "form-group",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "form-control ",
                  "onChange": [Function],
                  "placeholder": "Hasło",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
            "",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <button
                  className="btn btn-primary LoginButton"
                  type="submit"
>
                  Zaloguj
</button>,
                "className": "form-group",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Zaloguj",
                  "className": "btn btn-primary LoginButton",
                  "type": "submit",
                },
                "ref": null,
                "rendered": "Zaloguj",
                "type": "button",
              },
              "type": "div",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="btn btn-info"
              onClick={[Function]}
>
              Zarejestruj sie
</button>,
            "className": "form-group",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Zarejestruj sie",
              "className": "btn btn-info",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Zarejestruj sie",
            "type": "button",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
